### PR TITLE
br: support azure blob storage

### LINKED ISSUE #30583: br: support azure blob storage
## Feature Request

br supports azure blob storage.

### PR DESCRIPTION
**Pull request by Leavrth**
### What problem does this PR solve?

Issue Number: close #30583 

Problem Summary: support azure blob storage

### What is changed and how it works?

use azure blob storage as BR's external storage backend.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)

### Release note

```release-note
None
```

### CODE REVIEW

File: br/pkg/storage/azblob.go [hunk 1]
```diff
@@ -21,6 +22,59 @@ import (
...
+	azblobStorageClassOption = "azblob.storage-class"
+	azblobAccountName        = "azblob.account-name"
+	azblobAccountKey         = "azblob.account-key"
+)
+
+type AzblobBackendOptions struct {
+	Endpoint     string `json:"endpoint" toml:"endpoint"`
+	AccountName  string `json:"account_name" toml:"account_name"`
+	AccountKey   string `json:"account_key" toml:"account_key"`
+	StorageClass string `json:"storage_class" toml:"storage_class"`
```
Comment (kennytm) [inline br/pkg/storage/azblob.go]: the azure term is "access tier" not "storage class" 🤔

File: br/pkg/storage/azblob.go [hunk 2]
```diff
@@ -0,0 +1,508 @@
+package storage
```
Comment (3pointer) [inline br/pkg/storage/azblob.go:1]: need copyright.

File: br/pkg/storage/azblob_test.go [hunk 3]
```diff
@@ -0,0 +1,306 @@
+package storage
```
Comment (3pointer) [inline br/pkg/storage/azblob_test.go:1]: ditto

File: br/pkg/storage/azblob_test.go [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+	return azblob.NewServiceClientFromConnectionString(connStr, nil)
+}
+
+func (b *sharedKeyAzuriteClientBuilder) GetAccountName() string {
+	return "devstoreaccount1"
+}
+
+func checkAzuriteRunning() bool {
+	checkStatement := fmt.Sprintf("lsof -i:%d ", 10000)
+	output, _ := exec.Command("sh", "-c", checkStatement).CombinedOutput()
```
Comment (kennytm) [inline br/pkg/storage/azblob_test.go:33]: perhaps better to try if connecting to port 10000 succeeds.

File: br/pkg/storage/azblob.go [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+				containerExist = true
+				break RESPITER
+			}
+		}
+	}
+
+	respIter.PageResponse().RawResponse.Body.Close()
+	containerClient := serviceClient.NewContainerClient(options.Bucket)
+	if !containerExist {
+		_, err := containerClient.Create(ctx, nil)
```
Comment (3pointer) [inline br/pkg/storage/azblob.go:252]: according to document. 
```
Create creates a new container within a storage account. If a container with the same name already exists, the operation fails. For more information, see https://docs.microsoft.com/rest/api/storageservices/create-container.
```

so may be we can try create before, if failed then ignore create error?

File: br/pkg/storage/azblob.go [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+		return errors.Annotatef(err, "Failed to write azure blob file, file info: bucket(container)='%s', key='%s'", s.options.Bucket, s.withPrefix(name))
+	}
+	return nil
+}
+
+func (s *AzureBlobStorage) ReadFile(ctx context.Context, name string) ([]byte, error) {
+	client := s.containerClient.NewBlockBlobClient(s.withPrefix(name))
+	resp, err := client.Download(ctx, nil)
+	if err != nil {
+		return nil, errors.Annotatef(err, "Failed to read azure blob file, file info: bucket(container)='%s', key='%s'", s.options.Bucket, s.withPrefix(name))
```
Comment (3pointer) [inline br/pkg/storage/azblob.go:276]:
```suggestion
		return nil, errors.Annotatef(err, "Failed to download azure blob file, file info: bucket(container)='%s', key='%s'", s.options.Bucket, s.withPrefix(name))
```

**Review by 3pointer — APPROVED** [review]
rest lgtm

File: br/tests/br_azblob/_run.sh [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        echo "TEST: [$TEST_NAME] fail on database $DB${i}"
+    fi
+    echo "database $DB${i} [original] row count: ${row_count_ori[i]}, [after br] row count: ${row_count_new[i]}"
+done
+
+if $fail; then
+    echo "TEST: [$TEST_NAME] failed!"
+    exit 1
+else
+    echo "TEST: [$TEST_NAME] successd!"
```
Comment (kennytm) [inline br/tests/br_azblob/_run.sh:78]:
```suggestion
    echo "TEST: [$TEST_NAME] succeed!"
```

File: br/pkg/storage/azblob_test.go (hunk 4 above)
Comment (Leavrth) [reply br/pkg/storage/azblob_test.go:34]: Done. Check whether the error is connection error. PTAL

**Review by kennytm — APPROVED** [review]

**Pull request merged**